exla/lib/exla/mlir/value.ex

```diff
@@ -434,206 +542,414 @@ defmodule EXLA.MLIR.Value do
+    op(func, "stablehlo.dynamic_update_slice", [operand, updates] ++ starts, result_types) |> hd()
   end
 
   def reduce(
-        %Function{ref: reducer},
+        %Region{ref: reducer},
         [%Value{function: func} | _] = init_values,
         [%Value{function: func} | _] = inputs,
-        dimensions
+        dimensions,
+        typespecs
       ) do
-    init_value_refs = Enum.map(init_values, & &1.ref)
-    input_refs = Enum.map(inputs, & &1.ref)
-
-    refs =
-      EXLA.NIF.mlir_reduce(func.ref, reducer, init_value_refs, input_refs, dimensions)
-      |> unwrap!()
-
-    Enum.map(refs, &%Value{ref: &1, function: func})
+    operands = inputs ++ init_values
+    result_types = typespecs_to_mlir_types(typespecs)
+    attributes = [dimensions: attr_dense_i64_elements(dimensions)]
+    regions = [reducer]
+    op(func, "stablehlo.reduce", operands, result_types, attributes: attributes, regions: regions)
   end
 
   def window_reduce(
-        %Function{ref: reducer},
+        %Region{ref: reducer},
         [%Value{function: func} | _] = init_values,
         [%Value{function: func} | _] = inputs,
         window_dimensions,
         window_strides,
         input_dilations,
         window_dilations,
-        padding
+        padding,
+        typespecs
       ) do
-    init_value_refs = Enum.map(init_values, & &1.ref)
-    input_refs = Enum.map(inputs, & &1.ref)
+    operands = inputs ++ init_values
+    result_types = typespecs_to_mlir_types(typespecs)
 
-    refs =
-      EXLA.NIF.mlir_window_reduce(
-        func.ref,
-        reducer,
-        init_value_refs,
-        input_refs,
-        window_dimensions,
-        window_strides,
-        input_dilations,
-        window_dilations,
-        padding
-      )
-      |> unwrap!()
+    attributes = [
+      window_dimensions: attr_dense_i64_elements(window_dimensions),
+      window_strides: attr_dense_i64_elements(window_strides),
+      base_dilations: attr_dense_i64_elements(input_dilations),
+      window_dilations: attr_dense_i64_elements(window_dilations),
+      padding: attr_padding(padding)
+    ]
+
+    regions = [reducer]
 
-    Enum.map(refs, &%Value{ref: &1, function: func})
+    op(func, "stablehlo.reduce_window", operands, result_types,
+      attributes: attributes,
+      regions: regions
+    )
   end
 
   def map(
-        %Function{ref: mapper},
+        %Region{ref: mapper},
         [%Value{function: func} | _] = inputs,
-        dimensions
+        dimensions,
+        typespec
       ) do
-    input_refs = Enum.map(inputs, & &1.ref)
+    result_types = typespecs_to_mlir_types([typespec])
 
-    ref =
-      EXLA.NIF.mlir_map(func.ref, mapper, input_refs, dimensions)
-      |> unwrap!()
+    attributes = [
+      dimensions: attr_dense_i64_elements(dimensions)
+    ]
 
-    %Value{ref: ref, function: func}
+    regions = [mapper]
+
+    op(func, "stablehlo.map", inputs, result_types, attributes: attributes, regions: regions)
+    |> hd()
```

Comment:
Am I right that most of these `|> hd()` should only return a single value?

If so, I'd rather use hard-matching `[result] = ...` for unwrapping `op`.
Or at least, a `|> single_result!()` `defp` that does this, for more assertive code

Reply:
Good call, will change it to `|> one!()`!